mne/preprocessing/annotate_amplitude.py

```diff
@@ -0,0 +1,227 @@
+        (PTP). Valid **keys** can be any channel type present in the object.
+        The **values** are floats that set the minimum acceptable PTP. If the
+        PTP is smaller than this threshold, the segment will be annotated.
+        If float, the minimum acceptable PTP is applied to all channels.
+    bad_percent : float
+        The percentage of the time a channel can be bad.
+        Below this percentage, temporal bad marking (:class:`~mne.Annotations`)
+        will be used. Above this percentage, spatial bad marking
+        (:class:`info['bads'] <mne.Info>`) will be used.
+        Defaults to 5 (5%%).
+    min_duration : float
+        The minimum duration (sec) to consider as above or below threshold.
+        For some systems with low bit data representations, adjacent
+        time samples with exactly the same value are not totally uncommon.
+        Defaults to 0.005 (5 ms).
+    %(picks_good_data)s
+    %(verbose)s
+
+    Returns
+    -------
+    annot : instance of Annotations
+        The annotated bad segments.
+    bads : list
+        The channels detected as bad.
+
+    Notes
+    -----
+    This function may perform much faster if data are loaded in memory, as it
+    loads data one channel at a time (across all time points), which is
+    typically not an efficient way to read raw data from disk.
+
+    .. versionadded:: 1.0
+    """
+    _validate_type(raw, BaseRaw, 'raw')
+    picks_ = _picks_to_idx(raw.info, picks, 'data_or_ica', exclude='bads')
+    peak = _check_ptp(peak, 'peak', raw.info, picks_)
+    flat = _check_ptp(flat, 'flat', raw.info, picks_)
+    if peak is None and flat is None:
+        raise ValueError(
+            "At least one of the arguments 'peak' or 'flat' must not be None.")
+    bad_percent = _check_bad_percent(bad_percent)
+    min_duration = _check_min_duration(min_duration, raw.times[-1])
+    min_duration_samples = int(np.round(min_duration * raw.info['sfreq']))
+    bads = list()
+
+    # grouping picks by channel types to avoid operating on each channel
+    # individually
+    picks = {
+        ch_type: np.intersect1d(picks_of_type, picks_, assume_unique=True)
+        for ch_type, picks_of_type in _picks_by_type(raw.info, exclude='bads')
+        }
+    del picks_  # re-using this variable name below
+
+    any_flat = np.zeros(len(raw.times), bool)
+    any_peak = np.zeros(len(raw.times), bool)
+
+    # look for discrete difference above or below thresholds
+    logger.info('Finding segments below or above PTP threshold.')
+    for ch_type, picks_ in picks.items():
+        diff = np.abs(np.diff(raw._data[picks_, :], axis=1))
+
+        if flat is not None:
+            flat_ = diff <= flat[ch_type]
+            # reject too short segments
+            for start, stop in _2dim_mask_to_onsets_offsets(flat_):
+                if stop - start < min_duration_samples:
+                    flat_[start:stop] = False
```

Comment:
For my future self, this is not going to work, `flat_` is a 2D array `(n_channels, n_samples)`..